### PR TITLE
Add expired transactions to the mempool rejected list

### LINKED ISSUE #2848: Add expired transactions to the mempool rejected list
## Motivation

When a mempool transaction expires, we remove it from the mempool. But we should also add it to the rejected list, to avoid redownloading and reverifying it if it's sent again.

### Designs

At first the tx would fit into the `chain_rejected_same_effects` list which is used by txs rejected by excess capacity. But we should create a separate list for expired transactions. See [discussion](https://discord.com/channels/676527656170160146/716086297210650634/895785772291076117)

### PR DESCRIPTION
**Pull request by conradoplg**
## Motivation

When a mempool transaction expires, we remove it from the mempool. But we should also add it to the rejected list, to avoid redownloading and reverifying it if it's sent again.

## Solution

- Split `chain_rejected_same_effects` into two lists, to keep evicted and reject txs apart. (This was done by adding the lists to a HashMap, keyed by their respective error cases. This improved type safety a bit compared to just creating two separate lists)
- Add rejected txs to the respective list.

Closes https://github.com/ZcashFoundation/zebra/issues/2848

## Review

Anyone can review this.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

### CODE REVIEW

File: zebrad/src/components/inbound/tests.rs
Comment (dconnolly) [inline zebrad/src/components/inbound/tests.rs:412]:
```suggestion

    let queued_responses = match response {
```

File: zebrad/src/components/inbound/tests.rs
Comment (dconnolly) [inline zebrad/src/components/inbound/tests.rs:417]:
```suggestion

    assert_eq!(queued_responses.len(), 1);
```

File: zebrad/src/components/mempool.rs
Comment (teor2345) [inline zebrad/src/components/mempool.rs]: Confusing comment: we're rejecting by mined ID, but the comment says we need the unmined ID

```suggestion
    // we need a separate set, since reject() takes the original unmined ID,
    // then extracts the mined ID out of it
```

File: zebrad/src/components/mempool/storage.rs
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs:115]: This is a clever refactor - it saves us having a separate field for each error.

File: zebrad/src/components/mempool/storage.rs
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs:282]: Doc update: note that transactions in multiple lists are counted multiple times.

File: zebrad/src/components/mempool/storage.rs
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs:323]: Doc update:
* returns the error (not `true`) - I missed this update in a previous PR
* returns an arbitrary error if the transaction is in multiple lists

File: zebrad/src/components/mempool/storage.rs
Comment (teor2345) [inline zebrad/src/components/mempool/storage.rs]: Optional refactor: should we just call `rejection_error().is_some()` here?

(`HashMap::contains_key` just calls `get_inner`, so the performance should be similar to using `contains_key` directly.)

File: zebrad/src/components/inbound/tests.rs
Comment (teor2345) [inline zebrad/src/components/inbound/tests.rs:423]: Nice tests!

File: zebrad/src/components/mempool/storage.rs
Comment (teor2345) [reply zebrad/src/components/mempool/storage.rs:115]: Note: a transaction can be in multiple lists, but that is already the case due to `tip_rejected_exact` and `tip_rejected_same_effects`. That is ok, as long as the number of lists and errors stays small.

If we tried to make sure each transaction was only in the highest-priority list, we would complicate the code. And we'd risk subtle bugs or security issues.

I've added some specific comments on methods where the docs need to change.

**Review by teor2345 — COMMENTED** [review]
This looks good, we could merge it as-is.

I've made some suggestions about comments that could change.
I also suggested a refactor to remove near-duplicate code.

**Comment by teor2345** [comment]
Please feel free to merge with or without these changes - no need to wait for my re-approval.

I think this PR might conflict with ticket #2780, so we should try to get it merged on Monday.
(@dconnolly should feel free to approve and merge.)

**Comment by teor2345** [comment]
@conradoplg commit 51e59aec7402204fa507ac55fb1fb9993b2f8456 in PR #2855 resolves a tricky conflict between this PR and PR #2826.

You might want to cherry-pick it here, after you review it.

File: zebrad/src/components/mempool/storage.rs
Comment (conradoplg) [reply zebrad/src/components/mempool/storage.rs]: Done in f935a5fa

File: zebrad/src/components/mempool/storage.rs
Comment (conradoplg) [reply zebrad/src/components/mempool/storage.rs:323]: Done in a3c9053

File: zebrad/src/components/mempool/storage.rs
Comment (conradoplg) [reply zebrad/src/components/mempool/storage.rs:282]: Done in 149eb57

**Review by conradoplg — COMMENTED** [review]
Addressed comments! @dconnolly can you take a look per @teor2345's comment?

**Review by teor2345 — APPROVED** [review]
Looks good, let's merge

File: zebrad/src/components/mempool/storage.rs
Comment (conradoplg) [reply zebrad/src/components/mempool/storage.rs]: @teor2345 there's this missing unresolved conversation

**Review by dconnolly — COMMENTED** [review]
👍

**Pull request merged**
